### PR TITLE
Improve 'generic param from outer item' error for `Self` and inside `static`/`const` items

### LINKED ISSUE #109596: Use of `Self::SOME_ASSOC_CONST` inside a function-local `const` calls `Self` a "generic parameter from outer function"
### Code

Playground: <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=01d9dd114b740509a3538ce91aa2bf6a>.
```Rust
#[repr(u8)]
pub enum Foo { A, B, C }

impl Foo {
    pub const COUNT: usize = 3;
    pub fn blah() {
        const _: () = assert!(Self::COUNT < (u8::MAX as usize));
    }
}
```

A slightly more realistic (but also slightly larger) example is here: <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=01d9dd114b740509a3538ce91aa2bf6a> (this is more or less code that I tried to emit from a macro).

### Current output

```Shell
error[E0401]: can't use generic parameters from outer function
 --> src/lib.rs:7:31
  |
4 | impl Foo {
  | ---- `Self` type implicitly declared here, by this `impl`
...
7 |         const _: () = assert!(Self::COUNT < (u8::MAX as usize));
  |                               ^^^^^^^^^^^
  |                               |
  |                               use of generic parameter from outer function
  |                               use a type here instead

For more information about this error, try `rustc --explain E0401`.
error: could not compile `playground` due to previous error
```

### Desired output

```Shell
error[...]: can't use `Self` type inside function-local `const`.
```

Or something. I don't have strong feelings.

It should probably suggest replacing `Self::COUNT` with `Foo::COUNT` though, since that's enough to fix this.

### Rationale and extra context

`Self` here is not a generic parameter from the outer function (it's the implementing type from the `impl` block, although I'm not really sure that's a particularly clear way to refer to it either). Either way, the current message is wrong and misleading.

### Anything else?

Happens on nightly (`rustc 1.70.0-nightly (8be3c2bda 2023-03-24)`) and on stable (1.68.1).

### LINKED ISSUE #119936: Compile or improve error message for static `Self`
### Code

```Rust
impl MyStruct {
    fn instance() -> &'static Self {
        static MYSTRUCT: OnceLock<Self> = OnceLock::new();
        MYSTRUCT.get_or_init(|| Self { msg: "Hello world!" })
    }
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error[E0401]: can't use generic parameters from outer item
  --> src/main.rs:10:35
   |
8  | impl MyStruct {
   | ---- `Self` type implicitly declared here, by this `impl`
9  |     fn instance() -> &'static Self {
10 |         static MYSTRUCT: OnceLock<Self> = OnceLock::new();
   |                                   ^^^^
   |                                   |
   |                                   use of generic parameter from outer item
   |                                   refer to the type directly here instead
```

### Desired output

```Shell
It should either compile, or have a different message
```

### Rationale and extra context

The message says `use of generic parameter from outer item - refer to the type directly here instead`
This makes you believe that `Self` is a generic type, but it isn't. `MyStruct` can also be known at compile time at doesn't have generics attached to it, so maybe it could even compile.

### Other cases

```Rust
Changing the `Self` to `MyStruct`, which is the same thing, will make the code compile.
```

### Rust Version

```Shell
$ rustc --version --verbose
rustc 1.75.0 (82e1608df 2023-12-21)
binary: rustc
commit-hash: 82e1608dfa6e0b5569232559e3d385fea5a93112
commit-date: 2023-12-21
host: x86_64-unknown-linux-musl
release: 1.75.0
LLVM version: 17.0.6
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #109596
Fixes #119936

### CODE REVIEW

File: compiler/rustc_resolve/src/late.rs [hunk 1]
```diff
@@ -181,7 +189,8 @@ pub(crate) enum RibKind<'a> {
     FnOrCoroutine,
 
     /// We passed through an item scope. Disallow upvars.
-    Item(HasGenericParams),
+    /// Note if this is a static/const to help with some confusing diagnostics
+    Item(HasGenericParams, IsStaticOrConst),
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/late.rs:193]: I wonder if we can just record the item's `DefKind` in the rib. Seems a bit more general than just recording if it's a static or const. We should have access to that at this point, probably?

File: compiler/rustc_resolve/messages.ftl [hunk 2]
```diff
@@ -114,17 +114,27 @@ resolve_forward_declared_generic_param =
...
+        *[false] generic parameters
+    } from outer item
+    .label = use of {$is_self ->
+        [true] `Self`
+        *[false] generic parameter
+    } from outer item
     .refer_to_type_directly = refer to the type directly here instead
     .suggestion = try introducing a local generic parameter here
 
+resolve_generic_params_from_outer_item_const = a `const` creates a new item
```
Comment (compiler-errors) [inline compiler/rustc_resolve/messages.ftl:128]:
```suggestion
resolve_generic_params_from_outer_item_const = a `const` is a separate item from the body that contains it
```

idk if we should say "body" here, but i think "is separate" is important here. also the verb "creates" seems a bit weird, a `const` *is* an item, it doesn't *create* an item, at least in my mental model.

Comment (clubby789) [reply compiler/rustc_resolve/messages.ftl:128]: "separate item from that which contains it", maybe?

Comment (compiler-errors) [reply compiler/rustc_resolve/messages.ftl:128]: "that which contains it" feels a bit circular 😵

if you don't like "body", maybe just "the item that contains it"

File: compiler/rustc_resolve/src/ident.rs [hunk 3]
```diff
@@ -1155,7 +1153,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             }
             Res::Def(DefKind::TyParam, _) | Res::SelfTyParam { .. } | Res::SelfTyAlias { .. } => {
                 for rib in ribs {
-                    let has_generic_params: HasGenericParams = match rib.kind {
+                    let (has_generic_params, defkind) = match rib.kind {
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/ident.rs:1156]: There are 544 usages of `def_kind` in the compiler and 8 usages of `defkind`. Can you rename it to `def_kind`?

File: compiler/rustc_resolve/src/late.rs [hunk 4]
```diff
@@ -2471,11 +2478,14 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
...
-                    } else {
-                        HasGenericParams::No
-                    }),
+                    RibKind::Item(
+                        if self.r.tcx.features().generic_const_items {
+                            HasGenericParams::Yes(generics.span)
+                        } else {
+                            HasGenericParams::No
+                        },
+                        DefKind::Const,
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/late.rs:2487]: Is it possible to call something like `self.r.def_kind(def_id)`? We should have the item's def kind at this point.... I don't think it'll lead to query cycles.

I think you can turn the node_id -> def_id.

**Comment by cjgillot** [comment]
I'd like to be prudent with the perf effect, as we thread a lot more data.
@bors try @rust-timer queue

**Comment by compiler-errors** [comment]
r? compiler-errors @bors r+ rollup=maybe

**Pull request merged**